Outreachy PDF Form generation

bjohare opened the pull request:
### Proposed changes in this pull request

#### PDF Form generation implemented as part of Outreachy
* Implements ideas at https://github.com/Cadasta/cadasta-project-ideas/issues/12
* Thanks to @kavindya89

This PR adds PDF Form generation from project questionnaires. It adds an additional `PDFForm` model and related view classes, mixins and tests to the `questionnaires` app. In addition it adds a `PDFGenerator` class which leverages `WeasyPrint` to do the actual PDF generation. WeasyPrint system dependencies have been added to the cadasta application provisioning configuration.

### When should this PR be merged

After implementation of https://github.com/Cadasta/cadasta-platform/issues/1414

### Risks

Introduces a migration to the `questionnaires` app but risks should be low.

### Follow-up actions

None required.

### Checklist (for reviewing)

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?** 
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

oliverroick commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -0,0 +1,50 @@
+            if question.question_group is None:
+                template_questions_list.append(question)
+
+        if questionnaire is not None:
```
This condition is not needed. If no questionnaire is found it would break in line 22 already.

bjohare replied:
Right..

oliverroick commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -0,0 +1,50 @@
+        questions_list = questionnaire.questions.all()
+        template_question_groups_list = questionnaire.question_groups.all()
+
+        for question in questions_list:
```
Couldn't we write this as a list comprehension?

bjohare replied:
indeed we could..

oliverroick commented on cadasta/questionnaires/forms.py:
```diff
@@ -0,0 +1,29 @@
+        self.questionnaire = questionnaire
+
+    def save(self, *args, **kwargs):
+        pdfform = super().save(commit=False, *args, **kwargs)
```
Having `pdfform` and `self.instance` here is confusing. I would just use `pdfform` and use that in line 23 to check if the instance has an ID.

bjohare replied:
Yup..

oliverroick commented on cadasta/questionnaires/views/mixins.py:
```diff
@@ -0,0 +1,90 @@
+            questionnaire = Questionnaire.objects.get(
+                        id=project.current_questionnaire)
+            context['questionnaire'] = questionnaire
+            context['has_questionnaire'] = True
```
Do we need to set `has_questionnaire` explicitly? Can't we use `questionnaire` itself in the template to check if the project has a questionnaire?

oliverroick commented on cadasta/questionnaires/views/mixins.py:
```diff
@@ -0,0 +1,90 @@
+class PDFFormViewMixin:
+    form_class = PDFFormCreateForm
+
+    def get_model_context(self):
```
I tend to say we should bake the functionality into `get_form_kwargs` because that's the only place where it's used. It's probably more readable that way.

clash99 commented:
@bjohare - I'm done with ui stuff on this one. Only thing worth noting is I have updated acceptable file text to include gif and tiff. Let me know if you see anything else that needs ui attention.  Thanks!

bjohare commented:
@oliverroick I'd appreciate a second review of this. I've added a `ResourceThumbnailMixin` to make thumbnail generation a bit more reusable. The `PDFForm` model now uses this to generate a thumbnail for the logo if one is provided. The PDF and UI templates now use the thumbnails. This is a bit of an interim measure since I guess at some point we will want to look at thumbnail generation again when we add support for user profile images. I've also added file-type validation for logo uploads.

oliverroick commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -43,7 +43,9 @@ def generate_pdf(self, absolute_uri):
-                                        'pdfform': self.pdfform})
+                                        'pdfform': self.pdfform,
+                                        'thumbnail_url':
+                                        self.pdfform.thumbnail})
```
You're passing `pdfform` into the template context already, couldn't you use `pdfform.thumbnail` in the template?

oliverroick commented on cadasta/resources/mixins.py:
```diff
@@ -37,3 +41,67 @@ def detach_object_resources(sender, instance, **kwargs):
+                        name + '-128x128.' + ext,
+                        open(write_path, 'rb').read())
+
+    def _get_mime(self):
```
This is bad, I think. `PDFForm` doesn't have a `mime_type` field, so every time the `_get_mime` is called, the file is downloaded from S3 opened and evaluated. If you create the thumbnail for the form logo, the file is downloaded from S3 twice. First, when you get mime_type and second when you create the thumbnail itself. 

I think we might need to store the mime_type or make this more efficient. I don't think we need that `mime_type` very often, only to create thumbnails for the logo really.

bjohare replied:
Ok, this can probably be wrapped into `create_thumbnails` itself so that its not opened twice. Maybe I can just check for an allowed set of file extensions in `create_thumbnails` and then only double-check the mime once the file is already on disk.. I'm not entirely convinced by this `ResourceThumbnailMixin` to be honest..

oliverroick commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -1,20 +1,27 @@
+        logger = logging.getLogger("weasyprint")
+        if not logger.handlers:
+            logger.addHandler(logging.NullHandler())
+        from weasyprint import HTML
```
Can we import this on top of the file?

bjohare replied:
Seem to remember this generating linting errors but I'll have another look..

bjohare replied:
I need to supress `weasyprint` logging before the `weasyprint` import but if I put this at the top I get linting errors.. `E402 module level import not at top of file`. If I dont suppress weasyprint output we'll have annoying warning messages about Pango versions when tests are run and running dev server.

oliverroick commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -24,22 +31,20 @@ def generate_pdf(self, absolute_uri):
+            for question in questions_list if question.question_group is None]
+
+        html_string = render_to_string('questionnaires/'
+                                       'pdf_form_generator.html',
```
Can we make this more readable with less indentation?

bjohare replied:
Sure..

oliverroick commented on cadasta/templates/questionnaires/modal_form_add_unauthorized.html:
```diff
@@ -0,0 +1,30 @@
+        </div>
+
+        <div class="modal-body">
+          <p>{% trans "No questionnaire has attached with the project" %}</p>
```
The wording here sounds a bit odd to me.

bjohare replied:
Good call.. I'll amend.

clash99 commented:
What is the level of permission you need to be able to print a form? I can't print it as a project manager (clash) or even a superuser (oroick) locally.  I'm getting a "You don't have permission to view this pdf form" message. I also think that if only certain roles are allowed to print the form, then we should hide the Forms button on the sidebar to others since the messaging about printing a pdf is confusing to receive upon selecting a "Form" button.

<img width="1323" alt="screenshot 2017-06-01 11 52 55" src="https://cloud.githubusercontent.com/assets/1119621/26691186/d4942e0e-46c1-11e7-8773-f7c1da66a052.png">

clash99 commented on cadasta/templates/organization/project_wrapper.html:
```diff
@@ -154,6 +154,12 @@ <h1 class="{% if is_administrator or is_allowed_add_location or is_allowed_add_r
         <span class="title">{% trans "Resources" %}</span>
       </a>
+    </li><!--
+    --><li class="forms">
```
Can we hide the subnav Forms tabs for users that aren't allowed to print pdfs?

bjohare replied:
@clash99 yeah, that makes sense, only project admins will see this..

clash99 commented on cadasta/templates/questionnaires/pdf_form_generator.html:
```diff
@@ -0,0 +1,299 @@
+
+<div class="printable-form">
+  <div class="media">
+    {% if thumbnail_url %}
```
This isn't displaying any image. I think we want the full image to display (vs the thumbnail). Previous code worked but not sure if there is a better way.
`    {% if pdfform.file.url is not None and pdfform.file.url != "" %}
      <div class="media-left">
        <img alt="Form Logo" src="{{ pdfform.file.url }}" style="width: 128px; margin-right: 20px;">
      </div>
    {% endif %}`

bjohare replied:
@clash99 we don't place a restriction on the dimensions of the uploaded image. As I understood it the image is intended as a logo which is why we're generating a thumbnail for it and displaying that in the pdf?

clash99 replied:
I have css there to restrict the width. I can add a max-height too if you think that will help but the way the thumbnail generated is cropped, I think that would be bad to display for something like a logo.

bjohare replied:
Ok, I'll add the original url for PDF generation but use the thumbnail in the UI..

clash99 commented:
The questions aren't being displaying correctly in the pdf. I created a new project using the [Urban Informal Settlements Form](http://localhost:8000/media/s3/uploads/xls-forms/qkcnqfacjybvgzj82sed35yi.xlsx) which has 11 location questions. When I create a form to print, I'm only getting 3 (2 of which are automatic). This was working correctly not too long ago so somewhere along the line it went awry.

clash99 commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -0,0 +1,50 @@
+        [template_questions_list.append(question)
+            for question in questions_list if question.question_group is None]
+
+        context = {
```
I suspect changes in this area is what broke the pdf from generating the questions.

clash99 commented on cadasta/questionnaires/download/generator.py:
```diff
@@ -0,0 +1,50 @@
+        context = {
+            'questionnaire': questionnaire,
+            'questions_list': template_questions_list,
+            'questions_groups_list': template_question_groups_list,
```
This line should be `'question_groups_list': template_question_groups_list,`.
That extra "s" is what is preventing the questionnaire questions from printing correctly.

bjohare replied:
fixed.

amplifi commented:
Needs refactoring to use the async pipeline instead of saving files to local server disk.

oliverroick commented:
Closing as this won't be integrated into the platform any more
